src/Analyser/MutatingScope.php

```diff
@@ -2891,34 +2892,41 @@ private function enterAnonymousFunctionWithoutReflection(
 				$nativeTypes[$paramExprString] = new MixedType();
 				continue;
 			}
+			$nonRefVariableNames[$variableName] = true;
 			if ($this->hasVariableType($variableName)->no()) {
 				$variableType = new ErrorType();
 			} else {
 				$variableType = $this->getVariableType($variableName);
 				$nativeTypes[$paramExprString] = $this->getNativeType($use->var);
 			}
 			$expressionTypes[$paramExprString] = ExpressionTypeHolder::createYes($variableType);
+		}
 
-			foreach ($this->expressionTypes as $exprString => $typeHolder) {
-				$expr = $this->exprStringToExpr((string) $exprString);
-				if ($expr === null) {
-					continue;
-				}
-				if ($expr instanceof Variable) {
-					continue;
-				}
-				$variable = (new NodeFinder())->findFirst([$expr], static fn (Node $node): bool => $node instanceof Variable);
+		foreach ($this->expressionTypes as $exprString => $typeHolder) {
+			$expr = $this->exprStringToExpr((string) $exprString);
+			if ($expr === null) {
+				continue;
+			}
+			if ($expr instanceof Variable) {
+				continue;
+			}
+			$variables = (new NodeFinder())->findInstanceOf([$expr], Variable::class);
+			if ($variables === []) {
+				continue;
```

Comment:
I think https://github.com/phpstan/phpstan-src/pull/1929 can be merged by adding `$experssionTypes[$exprString] = $typeHolder` into this line.